README.md
kdiff3-stibbons
===============

Clone of kdiff3 public repository http://sourceforge.net/p/kdiff3/code/ci/master/tree/kdiff3/


KDiff3-Readme
=============

Author: Joachim Eibl  (joachim.eibl at gmx.de)
Copyright: (C) 2002-2012 by Joachim Eibl
KDiff3-Version: 0.9.97

KDiff3 runs best on KDE but can be built without it, depending only on Qt-libs.
These are available for Un*x, Windows, Mac.
Thus there are many setup possibilities to consider.

From version 0.9.94 on the main target platform is KDE4/Qt4.
For KDiff3 for KDE3/Qt3 use KDiff3-0.9.92 or older.

Contents
--------

- Introduction
- License
- Requirements & Installation
  - For KDE4
  - With Qt4-libs, without KDE
    - for Linux/Un*x
    - for Windows
    - Debugging with MinGW under Windows:
    - for Mac: Building KDiff3 for Mac OSX (with Intel processor) as a
      universal binary
    - Creating and installing translation messages
- Additional hints


Introduction
------------

KDiff3 is a program that
- compares and merges two or three input files or directories,
- shows the differences line by line and character by character (!),
- provides an automatic merge-facility and
- an integrated editor for comfortable solving of merge-conflicts
- has support for KDE-KIO (ftp, sftp, http, fish, smb),
- has an intuitive graphical user interface,
- provides a context menu for KDE-Konqueror and Windows-Explorer,
- supports 64 bit systems. (Some build issues are discussed in here.)

Do you want help translating? Read the README in the po-subdirectory!


License
-------

    GNU GENERAL PUBLIC LICENSE, Version 2, June 1991
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Steet, Fifth Floor, Boston, MA 02111-1307
    USA

    For details see file "COPYING".


Requirements & Installation
---------------------------

- for KDE4:
   From version 0.9.94 on KDiff3 supports KDE4.
   For installation on most distributions you usually also need these packages
   (names as on opensuse):
   - g++ (g++, version 3.4 or newer)
   - libqt4-devel (Qt4-libs, version 4.4.0 or newer)
   - libkde4-devel (KDE4 Header files and development libraries)
   - libkonq-devel (optional, needed for the kdiff3-plugin, if not installed
                    then the contextmenu plugin for konqueror won't be built.)
   - cmake (>2.6, checks dependencies and creates the appropriate Makefiles for
     your system)

   Typically in a terminal (e.g. konsole) you cd into the kdiff3-directory and
   run `./configure kde4`. This script is essentially the same as these commands:

      mkdir releaseKde
      cd releaseKde
      # Find out your KDE4-directory and note the result
      kde4-config --prefix
      # create Makefile, replace <KDE4-prefix> with the prior result
      cmake .. -DCMAKE_INSTALL_PREFIX=<KDE4-prefix>
      # run make (compile and link)
      make
      # install as root user
      sudo make install

   Now KDiff3 should be ready to use. Type `kdiff3` to start it.
   There should also be a entry in your KDE-start menu in
   "(Applications->)Development".

   For creating a debug version:

      mkdir debug
      cd debug
      # Find out your KDE4-directory and note the result
      kde4-config --prefix
      cmake .. -DCMAKE_INSTALL_PREFIX=<KDE4-prefix> -DCMAKE_BUILD_TYPE=debugfull
      make
      sudo make install

   The `<KDE4-prefix>` depends on your distribution:
   The command `kde4-config --prefix` should tell you.
   - For opensuse 11 use `/usr`.

- Building KDiff3 with Qt4-libs, but without KDE
    - for Linux/Un*x
    - for Windows
    - for Mac

   The version 0.9.97 requires Qt 4.4.0 or newer (from http://qt-project.org/) for
   compilation.

   You always need:
      - `kdiff3-0.9.97.tar.gz`
      - for Un*x:    `gcc`, `g++` with version >=3.4.2
                     `Qt-X11-libraries`
      - for Windows: `Qt-win-libs`
                     If you have the Qt4 commercial edition, then compile with
                     Visual Studio.NET 2003 or newer.
                     If you use the Qt4 open source edition, then you'll need
                     MinGW.
                     I verified that compilation and debugging works with
                     these packages (probably newer ones will also work):
                     - Use `MinGW-5.1.4.exe` or newer to install the following
                       or newer:
                       - `gcc-core-3.4.5-20060117-1.tar.gz`
                       - `binutils-2.17.50-20060824-1.tar.gz`
                       - `gcc-g++-3.4.5-20060117-1.tar.gz`
                       - `mingw-runtime-3.14.tar.gz`
                       - `mingw32-make-3.81-2.tar.gz`
                       - `w32api-3.11.tar.gz`
                     - `gdb-6.3-2.exe` (for debugging)
      - for Mac:     `gcc`, `g++` with version >=3.4.2
                     Qt-Mac-libraries
      - Optional: Qt Creator IDE: Development environment optimized for Qt
        (from http://qt-project.org)

  Build-instructions (Un*x):
    Assuming Qt is correctly installed run "./configure qt4" which essentially
    does the same as:
    - `mkdir releaseQt`
    - `cd releaseQt`
    - `qmake ../src-QT4/kdiff3.pro`
    - `make` (or `gmake` for GNU-Make)

  Build-instructions (Windows):
    - If you use Qt Creator from the SDK just load the `src-QT4\kdiff3.pro` 
      file and build.
    On the command line (cmd.exe):
    - `set QTDIR=<QT4-dir>`         e.g.: `c:\qt\4.4.1`
      `set "PATH=%QTDIR%\bin;%PATH%`
    - Prepare compiler settings:
      For Visual Studio run the respective `vsvars32.bat`
      For MinGW add the path: E.g. set `PATH=c:\mingw\bin;%PATH%`
    - cd into the directory `kdiff3-0.9.97\src-Qt4` and type
    - `qmake kdiff3.pro`
    - Either `nmake` for Visual Studio or `mingw32-make` depending on what you
      have.  

  Debugging with MinGW under Windows:
    - Qt Creator worked very well as debugger frontend too!
    - Debugging without gdb on the console:
      - The `qt-win-opensource-4.4.1-mingw.exe` only installs release dlls. You
        will have to compile the debug dlls yourself. Enter the
        `qt-4.4.1-directory` and run `configure -debug` and then `mingw32-make`.
      - cd into the directory kdiff3-0.9.97\src-Qt4
      - edit the file `Makefile.Debug` and in the LFLAGS replace
        `-Wl,-subsystem,windows` with `-Wl,-subsystem,console` (this is
        necessary so that gdb can send a break signal to the running program)
      - `mingw32-make debug` (create a debuggable executable)
      - `gdb debug\kdiff3.exe`
      - At the "(gdb)"-prompt type `run` and enter to start.
      - While the program runs you can't set breakpoints, first interrupt it
        with Ctrl-C.
      - Use `help` to find out about how to use gdb. (Important commands: `run`,
        `break`, `backtrace`, `print`, `quit`)
         Using break with c++-methods requires to specify the parameter types
         too. Breakpoints in constructors might not work.

  Building KDiff3 (Qt4) for Mac OSX: 
    0) Build with Qt Creator from the SDK and src-QT4\kdiff3.pro.
    Building a universal binary (by Michael Schmidt, perhaps outdated):
    1) Install Qt/Mac 4.x with thread support, static libraries and universal
    flag. (For Qt4 run "configure -static"; only necessary if the final binary
    should not require shared Qt4-libs.)
    2) `cd .../kdiff3-0.9.97/src-QT4`
    3a) Create a makefile for kdiff3 by executing:
       `qmake kdiff3.pro`
    3b) optionally generate XCode-Project (for debugging purposes only) by
    executing:
       `qmake -spec macx-xcode kdiff3.pro`
    4) Execute:
       `make`
    5) After the build completes copy kdiff3 from the `kdiff3-0.9.97/src-QT4`
       directory to:
       `/Applications`

  Creating and installing the translation messages
    The po directory contains translations from the KDE-translation teams.
    If you use the Qt-only-version of KDiff3, then the installation described
    above won't install translations automatically.
    To create and install the translations:
    1) `cd .../kdiff3-0.9.97/po`
    2) `sh create_qm_files install`    (asks for a super user password)
    Just to create the files in the po directory (as needed for the windows
    version):
    2) `sh create_qm_files local`


(End of KDiff3 with Qt4-instructions)
------------------------------------------------------------------------

Additional hints
----------------

   Start from commandline:
   - Comparing 2 files:     `kdiff3 file1 file2`
   - Merging 2 files:       `kdiff3 file1 file2 -o outputfile`
   - Comparing 3 files:     `kdiff3 file1 file2 file3`
   - Merging 3 files:       `kdiff3 file1 file2 file3 -o outputfile`
        Note that file1 will be treated as base of file2 and file3.

   If all files have the same name but are in different directories, you can
   reduce typework by specifying the filename only for the first file. E.g.:
   - Comparing 3 files:     `kdiff3 dir1/filename dir2 dir3`
   (This also works in the open-dialog.)

   If you start without arguments, then a dialog will appear where you can
   select your files via a filebrowser.

   For more documentation, see the help-menu or the subdirectory doc.

   Have fun!


kdiff3/ChangeLog
Version 0.9.98 - 2014-07-03
===========================
- Text rendering now with QTextLayout fixes the following issues
  - Correct handling for variable width fonts.
  - Corrected display of highlighted text with Qt4.8.x on Ubuntu and Mac.
  - Improved handling of texts with both right to left and left to right languages (mixed Arabic and western texts).
  - Improved handling of Chinese and Japanese.
  - Whitespace characters are now shown as dots for spaces and arrows for tabs, and not only in differences.
  - Fixed symlink comparison (Qt4 symLinkTarget returns absolute paths) 
- Text analysis for rendering with QTextLayout is interruptable and multithreaded.
  (See progressbar and abort-button in statusbar)
- Fix for saving to relative path in KDE-environments. (Patch from Harald Sitter)
- Fixed bug in 0.9.97: Directory compare was always case sensitive.
- Fix for saving files on KDE with relative path specified via command line option -o.
- Fixed problem with KIO (nonlocal urls).
- Improved Mac support.
- Write --confighelp information to stdout instead of stderr.
- Directory Merge Window: Enabled state of "Delete A And B" now also depends on existence of source file A.
- Works now with Qt4 and Qt5
- Progress dialog during printing.
- Workaround for bug in QSplitter::childEvent that broke QFileDialog::getSaveFileName

Version 0.9.97 - 2012-08-10
===========================
- Memory usage optimized for comparison of large directories. (ca. 1/5 needed)
- On Windows use config file .kdiff3rc next to kdiff3.exe if exists.
- In overwiev for two way diff show if only one side contains text.
- Fix for Fedora by Neal Becker in src-QT4/kdiff3part.desktop.
- When word wrap is active toggling line numbers now recalculates the word wrap.
- Removed confusing "For compatibility ..." hint from option -qall
- Fixed mouse wheel problem. (Patch by David Hay)
- Change an encoding in diff text window via click on encoding label. (Patch by Alexey Kostromin)
- Fix for tab-key moving focus instead of adding a tab character in MergeResultWindow.
- Workaround for git on Cygwin that allows KDiff3 to find files in a Cygwin "/tmp" directory 
  when environment variable "CYGWIN_BIN" is set. (Patch by Nigel Stewart)
- Removed iostream dependency (Patch by Nigel Stewart)
- Regression test framework (by Maurice van der Pot)
- Documentation patch (by Burkard Lueck)
- Select text in Find dialog (by Eike Sauer)
- If text is selected in either input or output window use that in Find dialog.
- Added default directory anti patterns ".hg" and ".git" and file anti patterns ".rej" and ".bak" for better mercurial and git integration.
- Command line option --cs doesn't change the config value permanently anymore.
- On KDE: Not creating a KDiff3Part anymore.
- Windows (Vista or Win 7): Shell context menu in directory comparison view now again displays text.
- Windows 64 bit: Diff-Ext-for-KDiff3 context menu now installed as 32 bit and 64 bit versions to allow other 32 bit programs to use it too.
- Windows 64 bit specific installer.

Version 0.9.96 - 2011-09-02
===========================
- KDiff3FileItemActionPlugin : Context menu that also works in dolphin (for KDE>=4.6)
- Parser for preprocessor commands. (Allows single apostrophs ')
- On Windows if the preprocessor command is "sed" try to use a "bin\sed.exe" next to the kdiff3.exe, if available.
- Warn if conversion errors appear (Invalid characters)
- Fix crash on A/B-overview (infinite recursion)
- Fix clearcase temp files not deleted problem on windows
- KDiff3 plugin: When launching KDiff3 konqueror isn't blocked anymore
- String corrections (Frederik Schwarzer)
- Fixed writing to KIO.
- OS2-Port (Patch by Silvan Scherrer)
- Fixed problem where destination directory would be renamed or deleted during copy operation. 
  Now if the destination directory exists only the files inside will be copied.
- In merge: Separate lines where the automatic choice would be the same but for different reasons.
- Fixed some problems with huge files in directory comparison mode (>2GB) (but direct comparison is still not possible)
- Fixed documentation compile errors with KDE>=4.5
- Fixed white space merge default options
- Fixed regexp test tool.
- Exclude printing code if Qt was compiled without printing support.
- For Windows: Fixed handling of unicode characters in command line parameters.
- Improved "old mac" lineendstyle handling: Break lines.
- Detect encoding specified in xml header or html "meta" tag. 

Version 0.9.95 - 2009/03/03
===========================
- Show line end style for each file.
- Updated message translations.
- Fixed permissions when writing executable file. (Un*x only)
- Fixed IgnorableCmdLineOptions (important for SVNs '-u'-option)
- Directory merge: Error when either B or C is changed and the other is deleted. (User choice required.)
- Qt-only Un*x-version looks for translations in /usr/share/locale/<lang>/LC_MESSAGES/kdiff3.qm
  (for Debian, by Eike Sauer)
- New script: po/create_qm_files: To create and install translations for Qt-only version.

Version 0.9.94 - 2009/01/17
===========================
- Fix for hidden text windows with --auto-flag.
- Fix for pasting clipboard truncated text if it contained characters that needed more than one byte in UTF-8 encoding.
- Fix for horizontal scrolling if word wrap is enabled.
- Directory tree: files hidden due to options (e.g. patterns etc.) don't affect folder equality any more.
- KDE: KIO-progress dialog is now hidden. (KDiff3 has its own progress dialog.)
- Directory merge: Default op for change in either B or C and delete in the other is now merge (previously copy).
- Directory merge: Not preserving merge operation after reload, because it might have changed.

Version 0.9.93 - 2009/01/06
===========================
- Support for KDE4 (with much porting help from Valentin Rusu)
- Fix for diff_ext_for_kdiff3 (by Sergey Zorin)
- Win32-Installation: SendTo-integration fixed for Vista.

kdiff3/src-QT4/main.cpp
/***************************************************************************
                          main.cpp  -  Where everything starts.
                             -------------------
    begin                : Don Jul 11 12:31:29 CEST 2002
    copyright            : (C) 2002-2007 by Joachim Eibl
    email                : joachim.eibl at gmx.de
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
#include "stable.h"
#include <kcmdlineargs.h>
#include <kaboutdata.h>
#include <klocale.h>
#include "kdiff3_shell.h"
#include <kstandarddirs.h>
#include "version.h"
#include <QTextCodec>
#include <QFile>
#include <QTextStream>
#include <QTranslator>
#include <QLocale>
#include <QFont>
#include <QClipboard>
#include <QMainWindow>
#include <vector>

#ifdef KREPLACEMENTS_H
#include "optiondialog.h"
#endif
#include "common.h"

void initialiseCmdLineArgs( KCmdLineOptions& options )
{
   QString configFileName = KStandardDirs().findResource("config","kdiff3rc");
   QFile configFile( configFileName );
   QString ignorableOptionsLine = "-u;-query;-html;-abort";
   if ( configFile.open( QIODevice::ReadOnly ) )
   {
      QTextStream ts( &configFile );
      while(!ts.atEnd())
      {
         QString line = ts.readLine();
         if ( line.startsWith("IgnorableCmdLineOptions=") )
         {
            int pos = line.indexOf('=');
            if (pos>=0)
            {
               ignorableOptionsLine = line.mid(pos+1);
            }
            break;
         }
      }
   }

#ifdef KREPLACEMENTS_H
   QStringList sl = ignorableOptionsLine.split( '|' );
#else
   QStringList sl = ignorableOptionsLine.split( ',' );
#endif
   if (!sl.isEmpty())
   {
      QStringList ignorableOptions = sl.front().split( ';' );
      for (QStringList::iterator i=ignorableOptions.begin(); i!=ignorableOptions.end(); ++i)
      {
         (*i).remove('-');
         if (!(*i).isEmpty())
         {
            if ( i->length()==1 )
               options.add( i->toLatin1() ).add("ignore", ki18n("Ignored. (User defined.)") );
            else
               options.add( i->toLatin1(), ki18n("Ignored. (User defined.)") );
         }
      }
   }
}

#ifdef _WIN32
#include <process.h>
// This command checks the comm
static bool isOptionUsed(const QString& s, int argc, char* argv[])
{
   for(int j=0; j<argc; ++j )
   {
      if( QString("-"+s) == argv[j] || QString("--"+s)==argv[j] )
      {
         return true;
      }
   }
   return false;
}
#endif

class ContextFreeTranslator : public QTranslator
{
public:
   ContextFreeTranslator( QObject* pParent ) : QTranslator(pParent) {}
   QString translate(const char* context, const char* sourceText, const char* comment ) const
   {
      if ( context != 0 )
         return QTranslator::translate(0,sourceText,comment);
      else
         return QString();
   }
};

int main(int argc, char *argv[])
{
#ifdef _WIN32
   /* KDiff3 can be used as replacement for the text-diff and merge tool provided by
      Clearcase. This is experimental and so far has only been tested under Windows.

      There are two ways to use KDiff3 with clearcase
      -  The file lib/mgrs/map contains the list of compare/merge tasks on one side and 
         the tool on the other. Originally this contains only clearcase tools, but you can
         edit this file and put kdiff3 there instead. (Recommended method)
      -  Exchange the original program with KDiff3: (Hackish, no fine control)
         1. In the Clearcase "bin"-directory rename "cleardiffmrg.exe" to "cleardiffmrg_orig.exe".
         2. Copy kdiff3.exe into that "bin"-directory and rename it to "cleardiffmrg.exe".
            (Also copy the other files that are needed by KDiff3 there.)
         Now when a file comparison or merge is done by Clearcase then of course KDiff3 will be
         run instead.
         If the commandline contains the option "-directory" then KDiff3 can't do it but will
         run "cleardiffmrg_orig.exe" instead.
   */

   // Write all args into a temporary file. Uncomment this for debugging purposes.
   /*
   FILE* f = fopen(QDir::toNativeSeparators(QDir::homePath()+"//kdiff3_call_args.txt").toLatin1().data(),"w");
   for(int i=0; i< argc; ++i)
      fprintf(f,"Arg %d: %s\n", i, argv[i]);
   fclose(f);
   
   // Call orig cleardiffmrg.exe to see what result it returns.
   int result=0;
   result = ::_spawnvp(_P_WAIT , "C:\\Programme\\Rational\\ClearCase\\bin\\cleardiffmrg.exe", argv );
   fprintf(f,"Result: %d\n", result );
   fclose(f);
   return result;
   */

   // KDiff3 can replace cleardiffmrg from clearcase. But not all functions.
   if ( isOptionUsed( "directory", argc,argv ) )
   {
      return ::_spawnvp(_P_WAIT , "cleardiffmrg_orig", argv );      
   }

#endif
#ifdef Q_OS_OS2
   // expand wildcards on the command line
   _wildcard(&argc, &argv);
#endif

   //QApplication::setColorSpec( QApplication::ManyColor ); // Grab all 216 colors

   const QByteArray& appName = QByteArray("kdiff3");
   const QByteArray& appCatalog = appName;
   const KLocalizedString i18nName = ki18n("kdiff3");
   QByteArray appVersion = QByteArray( VERSION );
   if ( sizeof(void*)==8 )
      appVersion += " (64 bit)";
   else if ( sizeof(void*)==4 )
       appVersion += " (32 bit)";
   const KLocalizedString description = ki18n("Tool for Comparison and Merge of Files and Directories");
   const KLocalizedString copyright = ki18n("(c) 2002-2014 Joachim Eibl");
   const QByteArray& homePage = "http://kdiff3.sourceforge.net/";
   const QByteArray& bugsAddress = "joachim.eibl" "@" "gmx.de";
   KAboutData aboutData( appName, appCatalog, i18nName, 
         appVersion, description, KAboutData::License_GPL_V2, copyright, description, 
         homePage, bugsAddress );

   aboutData.addAuthor(ki18n("Joachim Eibl"), KLocalizedString(), QByteArray("joachim.eibl" "@" "gmx.de"));
   aboutData.addCredit(ki18n("Eike Sauer"), ki18n("Bugfixes, Debian package maintainer") );
   aboutData.addCredit(ki18n("Sebastien Fricker"), ki18n("Windows installer") );
   aboutData.addCredit(ki18n("Stephan Binner"), ki18n("i18n-help"), QByteArray("binner" "@" "kde.org") );
   aboutData.addCredit(ki18n("Stefan Partheymueller"), ki18n("Clipboard-patch" ));
   aboutData.addCredit(ki18n("David Faure"), ki18n("KIO-Help"), QByteArray("faure" "@" "kde.org" ));
   aboutData.addCredit(ki18n("Bernd Gehrmann"), ki18n("Class CvsIgnoreList from Cervisia" ));
   aboutData.addCredit(ki18n("Andre Woebbeking"), ki18n("Class StringMatcher" ));
   aboutData.addCredit(ki18n("Michael Denio"), ki18n("Directory Equality-Coloring patch"));
   aboutData.addCredit(ki18n("Manfred Koehler"), ki18n("Fix for slow startup on Windows"));
   aboutData.addCredit(ki18n("Sergey Zorin"), ki18n("Diff Ext for Windows"));
   aboutData.addCredit(ki18n("Paul Eggert, Mike Haertel, David Hayes, Richard Stallman, Len Tower"), ki18n("GNU-Diffutils"));
   aboutData.addCredit(ki18n("Tino Boellsterling, Timothy Mee"), ki18n("Intensive test, use and feedback"));
   aboutData.addCredit(ki18n("Michael Schmidt"), ki18n("Mac support"));
   aboutData.addCredit(ki18n("Valentin Rusu"), ki18n("KDE4 porting"), QByteArray("kde" "@" "rusu.info"));
   aboutData.addCredit(ki18n("Albert Astals Cid"), ki18n("KDE4 porting"), QByteArray("aacid" "@" "kde.org"));
   aboutData.addCredit(ki18n("Silvan Scherrer"), ki18n("OS2 port") );
   aboutData.addCredit(ki18n("Nigel Stewart"), ki18n("Cygwin support") );
   aboutData.addCredit(ki18n("Maurice van der Pot"), ki18n("Framework for automated tests") );

   aboutData.addCredit(ki18n("+ Many thanks to those who reported bugs and contributed ideas!"));

   KCmdLineArgs::init( argc, argv, &aboutData );

   KCmdLineOptions options;
   // ignorable command options
   options.add( "m" ).add( "merge", ki18n("Merge the input."));
   options.add( "b" ).add( "base file", ki18n("Explicit base file. For compatibility with certain tools.") );
   options.add( "o" ).add( "output file", ki18n("Output file. Implies -m. E.g.: -o newfile.txt"));
   options.add( "out file",    ki18n("Output file, again. (For compatibility with certain tools.)") );
   options.add( "auto",        ki18n("No GUI if all conflicts are auto-solvable. (Needs -o file)") );
   options.add( "qall",        ki18n("Don't solve conflicts automatically.") );
   options.add( "L1 alias1",   ki18n("Visible name replacement for input file 1 (base).") );
   options.add( "L2 alias2",   ki18n("Visible name replacement for input file 2.") );
   options.add( "L3 alias3",   ki18n("Visible name replacement for input file 3.") );
   options.add( "L" ).add( "fname alias", ki18n("Alternative visible name replacement. Supply this once for every input.") );
   options.add( "cs string",   ki18n("Override a config setting. Use once for every setting. E.g.: --cs \"AutoAdvance=1\"") );
   options.add( "confighelp",  ki18n("Show list of config settings and current values.") );
   options.add( "config file", ki18n("Use a different config file.") );

   // other command options
   options.add( "+[File1]", ki18n("file1 to open (base, if not specified via --base)") );
   options.add( "+[File2]", ki18n("file2 to open") );
   options.add( "+[File3]", ki18n("file3 to open") );

#if defined(_WIN32) || defined(Q_OS_OS2)
   // on Windows applicationDirPath() is called and needs a QApplication
   KApplication app;
   initialiseCmdLineArgs( options );
   KCmdLineArgs::addCmdLineOptions( options );
   app.parseOptions();
#else
   // on other systems options should be initialized before creating the QApplication 
   initialiseCmdLineArgs( options );
   KCmdLineArgs::addCmdLineOptions( options );
   KApplication app;
#endif

#ifdef KREPLACEMENTS_H
   QString locale;

   locale = app.config()->readEntry("Language", "Auto");
   int spacePos = locale.indexOf(' ');
   if (spacePos>0) locale = locale.left(spacePos);
   ContextFreeTranslator kdiff3Translator( 0 );
   QTranslator qtTranslator( 0 );
   if (locale != "en_orig")
   {
      if ( locale == "Auto" || locale.isEmpty() )
         locale = locale = QLocale::system().name().left(2);
         
      QString translationDir = getTranslationDir(locale);
      kdiff3Translator.load( QString("kdiff3_")+locale, translationDir );
      app.installTranslator( &kdiff3Translator );
      
      qtTranslator.load( QString("qt_")+locale, translationDir );
      app.installTranslator( &qtTranslator );
   }
#endif

#ifndef QT_NO_SESSIONMANAGER
  if (app.isSessionRestored())
  {
     RESTORE(KDiff3Shell);
  }
  else
#endif
  {
     KDiff3Shell* p = new KDiff3Shell();
     p->show();
     p->setWindowState( p->windowState() | Qt::WindowActive ); // Patch for ubuntu: window not active on startup
  }
//app.installEventFilter( new CFilter );
  int retVal = app.exec();
  if (QApplication::clipboard()->text().size() == 0)
     QApplication::clipboard()->clear(); // Patch for Ubuntu: Fix issue with Qt clipboard
  return retVal;
}

// Suppress warning with --enable-final
#undef VERSION
